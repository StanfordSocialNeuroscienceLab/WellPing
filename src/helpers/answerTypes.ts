import { AnswerEntity } from "../entities/AnswerEntity";
import { Question, QuestionsList, Choice } from "./types";

// We use `{ value: ... }` because it seems that `"simple-json"`
// does not support simply mixing normal values.
// This also increases the consistency of the data.

export type SliderAnswerData = {
  value: number;
};

export type ChoicesWithSingleAnswerAnswerData = {
  value: Choice;
};

// We use an array of tuples here (instead of `{ [key: string]: boolean }`)
// because the order shown on the screen can be kept in the results.
// This is potentially helpful for analysis later if the order is randomized
// as now we know the order the user saw the choices in.
export type ChoicesWithMultipleAnswersAnswerChoices = [Choice, boolean][];
export type ChoicesWithMultipleAnswersAnswerData = {
  value: ChoicesWithMultipleAnswersAnswerChoices;
};

export type YesNoAnswerData = {
  value: boolean;
};

export type MultipleTextAnswerData = {
  value: string[];
};

export type HowLongAgoAnswerDataType = [number | null, string | null];
export type HowLongAgoAnswerData = {
  value: HowLongAgoAnswerDataType;
};

export type AnswerData =
  | SliderAnswerData
  | ChoicesWithSingleAnswerAnswerData
  | ChoicesWithMultipleAnswersAnswerData
  | YesNoAnswerData
  | MultipleTextAnswerData
  | HowLongAgoAnswerData;

export interface AnswersList {
  [id: string]: AnswerEntity;
}

export interface QuestionScreenProps {
  question: Question;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: () => boolean) => void;
}
