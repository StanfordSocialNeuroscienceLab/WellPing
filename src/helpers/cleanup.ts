import { secureRemoveAllAnswersAsync } from "./secureStore/answer";
import { secureRemoveAllPingsAsync } from "./secureStore/ping";

export async function clearAllPingsAndAnswersAsync(): Promise<void> {
  await secureRemoveAllAnswersAsync();
  await secureRemoveAllPingsAsync();
}
