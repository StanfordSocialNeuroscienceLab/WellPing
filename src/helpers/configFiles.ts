import { isThisWeek } from "date-fns";

import { SurveyFile, Names, StudyID, StudyInfo, StreamName } from "./types";

export function getSurveyFile(): SurveyFile {
  const survey: SurveyFile = require("../../config/survey.json");
  return survey;
}

// TODO: DECOUPLE FUNCTIONS LIKE THIS
export function getNamesFile(): Names {
  const names: Names = require("../../config/names.json");
  return names;
}

export function getStudyInfo(): StudyInfo {
  return getSurveyFile().studyInfo;
}

export function getAllStreamNames(): StreamName[] {
  return Object.keys(getSurveyFile().meta.startingQuestionIds) as StreamName[];
}

export function getStudyId(): StudyID {
  return getStudyInfo().id;
}

export function isTimeThisWeek(time: Date): boolean {
  return isThisWeek(time, { weekStartsOn: getStudyInfo().weekStartsOn });
}
