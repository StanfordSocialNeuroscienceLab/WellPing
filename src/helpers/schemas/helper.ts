export const idRegexCheck = (val: string) => /^\w+$/.test(val);
export const idRegexErrorMessage = (name: string) =>
  `${name} can only include letters, numbers, and "_".`;
