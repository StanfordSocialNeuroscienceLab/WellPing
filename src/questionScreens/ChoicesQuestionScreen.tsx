import cloneDeep from "lodash/cloneDeep";
import React from "react";
import {
  View,
  Text,
  TouchableOpacity,
  FlatList,
  StyleSheet,
} from "react-native";

import {
  QuestionScreenProps,
  ChoicesWithMultipleAnswersAnswerChoices,
  YesNoAnswerData,
  ChoicesWithSingleAnswerAnswerData,
  ChoicesWithMultipleAnswersAnswerData,
} from "../helpers/answerTypes";
import { QuestionType, shuffle } from "../helpers/helpers";
import { ChoicesQuestion, YesNoQuestion, Choice } from "../helpers/types";

export interface ChoiceItemProps {
  id: string;
  title: string;
  selected: boolean;
  onSelect: (id: string) => void;
}
export function ChoiceItem({ id, title, selected, onSelect }: ChoiceItemProps) {
  return (
    <TouchableOpacity
      onPress={() => onSelect(id)}
      style={[
        styles.item,
        { backgroundColor: selected ? "#b3995d" : "#F9F6EF" },
      ]}
      accessibilityLabel={`select ${title}`}
    >
      <Text style={styles.title}>{title}</Text>
    </TouchableOpacity>
  );
}

interface ChoicesQuestionScreenProps extends QuestionScreenProps {
  question: ChoicesQuestion | YesNoQuestion;
}

function initAnswerData(
  choices: Choice[],
): ChoicesWithMultipleAnswersAnswerChoices {
  // https://stackoverflow.com/a/26265095/2603230
  const defaultAnswers: ChoicesWithMultipleAnswersAnswerChoices = choices.reduce(
    (map, choice) => {
      map[choice.key] = false;
      return map;
    },
    {} as ChoicesWithMultipleAnswersAnswerChoices,
  );
  return defaultAnswers;
}

enum ChoicesAnswerType {
  SINGLE_SELECTION, // ChoicesWithMultipleAnswersAnswerChoices
  MULTIPLE_SELECTION, // string
  YESNO, // boolean
}

const YESNO_CHOICES_YES_KEY = "yes";
const YESNO_CHOICES_NO_KEY = "no";
const YESNO_CHOICES: Choice[] = [
  { key: YESNO_CHOICES_YES_KEY, value: "Yes" },
  { key: YESNO_CHOICES_NO_KEY, value: "No" },
];

const ChoicesQuestionScreen: React.ElementType<ChoicesQuestionScreenProps> = ({
  question,
  onDataChange,
  pipeInExtraMetaData,
}) => {
  let answerType: ChoicesAnswerType;
  switch (question.type) {
    case QuestionType.ChoicesWithSingleAnswer:
      answerType = ChoicesAnswerType.SINGLE_SELECTION;
      break;

    case QuestionType.ChoicesWithMultipleAnswers:
      answerType = ChoicesAnswerType.MULTIPLE_SELECTION;
      break;

    case QuestionType.YesNo:
      answerType = ChoicesAnswerType.YESNO;
      break;

    default:
      throw new Error("Wrong QuestionType in ChoicesQuestionScreen");
  }

  let choices: Choice[];
  if (answerType === ChoicesAnswerType.YESNO) {
    choices = YESNO_CHOICES;
  } else {
    const cQ = question as ChoicesQuestion;
    choices = cQ.choices;
  }

  const listRef = React.useRef<FlatList<{ id: string; title: string }>>(null);

  const [selected, setSelected] = React.useState<
    ChoicesWithMultipleAnswersAnswerChoices
  >(initAnswerData(choices));

  type FlatListData = {
    id: string;
    title: string;
  }[];
  const [flatListData, setFlatListData] = React.useState<FlatListData>([]);

  React.useEffect(() => {
    // We have to use `useEffect(..., [])` here to ensure this only runs once.
    // If we don't use `useEffect`, each time the user update the state, the choices will be re-shuffled.
    let tempFlatListData = choices.map((choice) => ({
      id: choice.key,
      title: pipeInExtraMetaData(choice.value),
    }));
    if (answerType !== ChoicesAnswerType.YESNO) {
      const cQ = question as ChoicesQuestion;
      if (cQ.randomizeChoicesOrder) {
        const randomizeExceptForChoiceIds =
          cQ.randomizeExceptForChoiceIds || [];

        const flatListShuffleableData = tempFlatListData.filter(
          (value) => !randomizeExceptForChoiceIds.includes(value.id),
        );
        const flatListFixedDataAtLast = tempFlatListData.filter((value) =>
          randomizeExceptForChoiceIds.includes(value.id),
        );
        tempFlatListData = [
          ...shuffle(flatListShuffleableData),
          ...flatListFixedDataAtLast,
        ];
      }
    }
    setFlatListData(tempFlatListData);
  }, []);

  return (
    <View style={{ paddingTop: 20 }}>
      <FlatList
        data={flatListData}
        renderItem={({ item }) => (
          <ChoiceItem
            id={item.id}
            title={item.title}
            selected={selected[item.id]}
            onSelect={(id: string) => {
              let newSelected: ChoicesWithMultipleAnswersAnswerChoices;
              if (answerType === ChoicesAnswerType.MULTIPLE_SELECTION) {
                newSelected = cloneDeep(selected);
              } else {
                // Reset everything to `false` if it is a single-selection question.
                newSelected = initAnswerData(choices);
              }
              newSelected[id] = !newSelected[id];
              setSelected(newSelected);

              switch (answerType) {
                case ChoicesAnswerType.MULTIPLE_SELECTION:
                  onDataChange({
                    value: newSelected,
                  } as ChoicesWithMultipleAnswersAnswerData);
                  break;

                case ChoicesAnswerType.SINGLE_SELECTION:
                  // Single-selection question only need the selected key as the data.
                  onDataChange({
                    value: id,
                  } as ChoicesWithSingleAnswerAnswerData);
                  break;

                case ChoicesAnswerType.YESNO:
                  onDataChange({
                    value: id === YESNO_CHOICES_YES_KEY,
                  } as YesNoAnswerData);
                  break;
              }
            }}
          />
        )}
        keyExtractor={(item) => item.id}
        extraData={selected}
        ref={listRef}
        style={{
          borderWidth: 1,
          borderColor: "lightgray",
        }}
      />
      {answerType !== ChoicesAnswerType.YESNO && (
        <Text style={{ textAlign: "center" }}>
          You may need to scroll to see all options.
        </Text>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  item: {
    backgroundColor: "#f9c2ff",
    padding: 10,
    marginVertical: 8,
  },
  title: {
    fontSize: 20,
  },
});

export default ChoicesQuestionScreen;
