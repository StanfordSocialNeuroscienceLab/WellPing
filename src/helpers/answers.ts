import { AnswerData, Answer } from "./answerTypes";
import { getAnswersPingIdsQuestionIdsListAsync } from "./asyncStorage/answersPingIdsQuestionIdsList";
import { AnswerSchema } from "./schemas/Answer";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
} from "./secureStore/answer";
import { Question, Ping } from "./types";

export async function getAnswersAsync(): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}
