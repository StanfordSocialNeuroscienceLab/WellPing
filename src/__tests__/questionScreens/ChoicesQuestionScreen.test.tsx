import React from "react";
import { Alert } from "react-native";
import {
  render,
  fireEvent,
  A11yAPI,
  FireEventAPI,
  waitFor,
  RenderAPI,
} from "react-native-testing-library";
import { ReactTestInstance } from "react-test-renderer";

import {
  YesNoAnswerData,
  ChoicesWithSingleAnswerAnswerData,
  ChoicesWithMultipleAnswersAnswerData,
} from "../../helpers/answerTypes";
import { QuestionType } from "../../helpers/helpers";
import {
  YesNoQuestion,
  ChoicesWithSingleAnswerQuestion,
  ChoicesWithMultipleAnswersQuestion,
  Choice,
} from "../../helpers/types";
import ChoicesQuestionScreen from "../../questionScreens/ChoicesQuestionScreen";
import { simplePipeInExtraMetaData } from "../helper";

export const getSelectionA11YLabel = (option: string) => `select ${option}`;
export const getSelection = (
  option: string,
  getAllByA11yLabel: A11yAPI["getAllByA11yLabel"],
) => {
  const selections = getAllByA11yLabel(getSelectionA11YLabel(option));
  expect(selections).toHaveLength(1);
  const selection = selections[0];
  return selection;
};

test.each([true, false])("YesNo question with input %p", async (inputValue) => {
  const question = {
    id: "YesNo",
    type: QuestionType.YesNo,
    question: "Are you happy?",
    next: null,
  } as YesNoQuestion;

  const {
    renderResults: { getAllByA11yLabel },
    mockOnDataChangeFn,
  } = await basicTestForChoicesQuestionScreenAsync(question);

  if (inputValue === true) {
    const yesSelection = getSelection(`Yes`, getAllByA11yLabel);
    fireEvent.press(yesSelection);

    expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(1, {
      value: true,
    } as YesNoAnswerData);
    expect(mockOnDataChangeFn).toHaveBeenCalledTimes(1);
  } else if (inputValue === false) {
    const noSelection = getSelection(`No`, getAllByA11yLabel);
    fireEvent.press(noSelection);

    expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(1, {
      value: false,
    } as YesNoAnswerData);
  }
});

const basicTestForChoicesQuestionScreenAsync = async (
  question:
    | YesNoQuestion
    | ChoicesWithSingleAnswerQuestion
    | ChoicesWithMultipleAnswersQuestion,
) => {
  let choices: Choice[];
  if (question.type === QuestionType.YesNo) {
    choices = [
      { key: "yes", value: "Yes" },
      { key: "no", value: "No" },
    ];
  } else {
    choices = question.choices;
  }

  // To make sure we can deterministically confirm the randomness.
  // https://stackoverflow.com/a/57730344/2603230
  const mathRandomSpy = jest
    .spyOn(global.Math, "random")
    .mockReturnValue(0.3141592653589793);

  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const renderResults = render(
    <ChoicesQuestionScreen
      key={question.id}
      question={question}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{}}
      allQuestions={{ [question.id]: question }}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );
  const { findAllByA11yLabel } = renderResults;

  expect(mockPipeInExtraMetaData).toHaveBeenCalledTimes(choices.length); // For each choices

  // There shouldn't need to be any validation
  expect(mockSetDataValidationFunction).not.toHaveBeenCalled();

  const selections = await findAllByA11yLabel(/select (.*?)/);
  expect(selections).toHaveLength(choices.length);

  const displayedList: string[] = selections.map((selection) =>
    selection.props.accessibilityLabel.replace(/^(select )/, ""),
  );
  if (question.type !== QuestionType.YesNo) {
    if (question.randomizeChoicesOrder) {
      expect(displayedList).not.toStrictEqual(FLATTENED_CHOICES_VALUES);
      if (question.randomizeExceptForChoiceIds) {
        // Sort the `randomizeExceptForChoiceIds` by the order of `choices`.
        const sortedRandomizeExceptForChoiceIds = [];
        for (const choice of choices) {
          if (question.randomizeExceptForChoiceIds.includes(choice.key)) {
            sortedRandomizeExceptForChoiceIds.push(choice.key);
          }
        }
        // Just a sanity check to make sure the items are the same.
        expect(sortedRandomizeExceptForChoiceIds).toHaveLength(
          question.randomizeExceptForChoiceIds.length,
        );

        for (let i = sortedRandomizeExceptForChoiceIds.length; i > 0; i--) {
          const key =
            sortedRandomizeExceptForChoiceIds[
              sortedRandomizeExceptForChoiceIds.length - i
            ];
          const value = question.choices.find((choice) => choice.key === key)
            ?.value;
          expect(displayedList[displayedList.length - i]).toBe(value);
        }
      }
    } else {
      expect(displayedList).toStrictEqual(FLATTENED_CHOICES_VALUES);
    }
  }

  expect(displayedList).toMatchSnapshot("displayed list");

  mathRandomSpy.mockRestore();

  return { renderResults, mockOnDataChangeFn };
};

const inputTestForChoicesWithSingleAnswerQuestionAsync = async (
  renderResults: RenderAPI,
  mockOnDataChangeFn: jest.Mock<any, any>,
  question: ChoicesWithSingleAnswerQuestion,
  inputValueSequence: string[],
) => {
  const { getAllByA11yLabel } = renderResults;

  let expectedResults: string | null = null;

  let calledTimes = 0;
  for (let i = 0; i < inputValueSequence.length; i++) {
    const inputValue = inputValueSequence[i];
    const selection = getSelection(inputValue, getAllByA11yLabel);
    fireEvent.press(selection);

    expectedResults = question.choices.find(
      (choice) => choice.value === inputValue,
    )?.key!;

    calledTimes += 1;
    expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(calledTimes, {
      value: expectedResults,
    } as ChoicesWithSingleAnswerAnswerData);
    expect(mockOnDataChangeFn).toHaveBeenCalledTimes(calledTimes);
  }

  if (expectedResults != null) {
    expect(expectedResults).toMatchSnapshot("data");
  }

  return renderResults;
};

const inputTestForChoicesWithMultipleAnswersQuestionAsync = async (
  renderResults: RenderAPI,
  mockOnDataChangeFn: jest.Mock<any, any>,
  question: ChoicesWithMultipleAnswersQuestion,
  inputValueSequence: string[],
) => {
  const { getAllByA11yLabel } = renderResults;

  const expectedResults = question.choices.reduce((map, choice) => {
    map[choice.key] = false;
    return map;
  }, {} as { [key: string]: boolean });

  let calledTimes = 0;
  for (let i = 0; i < inputValueSequence.length; i++) {
    const inputValue = inputValueSequence[i];
    const selection = getSelection(inputValue, getAllByA11yLabel);
    fireEvent.press(selection);

    const key = question.choices.find((choice) => choice.value === inputValue)
      ?.key!;
    expectedResults[key] = !expectedResults[key];

    calledTimes += 1;
    expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(calledTimes, {
      value: expectedResults,
    } as ChoicesWithMultipleAnswersAnswerData);
    expect(mockOnDataChangeFn).toHaveBeenCalledTimes(calledTimes);
  }

  expect(expectedResults).toMatchSnapshot("data");

  return renderResults;
};

const CHOICES = [
  { key: "wolf", value: "Wolf" },
  { key: "fox", value: "Fox" },
  { key: "coyote", value: "Coyote" },
  { key: "lynx", value: "Lynx" },
  { key: "panda", value: "Panda" },
  { key: "other", value: "Other" },
  { key: "idk", value: "I don't know" },
];
const FLATTENED_CHOICES_VALUES = CHOICES.map((choice) => choice.value);
const CHOICES_TEST_TABLE = [
  [["Wolf", "Coyote"], false, undefined],
  [["Lynx", "Panda", "Other"], true, ["other"]],
  [["Fox"], true, ["other", "idk"]],
  [
    ["Coyote", "Lynx", "Wolf", "Fox", "Coyote", "I don't know"],
    true,
    ["wolf", "fox", "other", "idk"],
  ],
  [
    ["Coyote", "Coyote", "Wolf", "Wolf", "Coyote"],
    true,
    ["idk", "other"], // The order should follow `choices`, not here.
  ],
  [
    [
      "Coyote",
      "Wolf",
      "Fox",
      "Lynx",
      "Panda",
      "Other",
      "Coyote",
      "Wolf",
      "Fox",
      "Lynx",
      "Panda",
      "Other",
      "I don't know",
    ],
    true,
    undefined,
  ],
  [[], false, undefined],
] as [string[], boolean, string[] | undefined][];
test.each(CHOICES_TEST_TABLE)(
  "ChoicesWithSingleAnswer question with input sequence %p (randomize order %p except for %p)",
  async (
    inputValueSequence,
    randomizeChoicesOrder,
    randomizeExceptForChoiceIds,
  ) => {
    const question = {
      id: "AnimalTest",
      type: QuestionType.ChoicesWithSingleAnswer,
      question: "What are you?",
      randomizeChoicesOrder,
      randomizeExceptForChoiceIds,
      choices: CHOICES,
      next: null,
    } as ChoicesWithSingleAnswerQuestion;

    const {
      renderResults,
      mockOnDataChangeFn,
    } = await basicTestForChoicesQuestionScreenAsync(question);

    await inputTestForChoicesWithSingleAnswerQuestionAsync(
      renderResults,
      mockOnDataChangeFn,
      question,
      inputValueSequence,
    );
  },
);

test.each(CHOICES_TEST_TABLE)(
  "ChoicesWithMultipleAnswers question with input sequence %p (randomize order %p except for %p)",
  async (
    inputValueSequence,
    randomizeChoicesOrder,
    randomizeExceptForChoiceIds,
  ) => {
    const question = {
      id: "AnimalTest",
      type: QuestionType.ChoicesWithMultipleAnswers,
      question: "What animals do you love?",
      randomizeChoicesOrder,
      randomizeExceptForChoiceIds,
      choices: CHOICES,
      next: null,
    } as ChoicesWithMultipleAnswersQuestion;

    const {
      renderResults,
      mockOnDataChangeFn,
    } = await basicTestForChoicesQuestionScreenAsync(question);

    await inputTestForChoicesWithMultipleAnswersQuestionAsync(
      renderResults,
      mockOnDataChangeFn,
      question,
      inputValueSequence,
    );
  },
);
