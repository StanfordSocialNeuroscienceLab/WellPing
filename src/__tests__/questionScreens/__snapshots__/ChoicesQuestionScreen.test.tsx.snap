// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChoicesWithMultipleAnswers question with input sequence ["Coyote", "Coyote", "Wolf", "Wolf", "Coyote"] (randomize order true except for ["I don't know", "Other"]): data 1`] = `
Array [
  Array [
    "Lynx",
    false,
  ],
  Array [
    "Coyote",
    true,
  ],
  Array [
    "Wolf",
    false,
  ],
  Array [
    "Panda",
    false,
  ],
  Array [
    "Fox",
    false,
  ],
  Array [
    "Other",
    false,
  ],
  Array [
    "I don't know",
    false,
  ],
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Coyote", "Coyote", "Wolf", "Wolf", "Coyote"] (randomize order true except for ["I don't know", "Other"]): displayed list 1`] = `
Array [
  "Lynx",
  "Coyote",
  "Wolf",
  "Panda",
  "Fox",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Coyote", "Lynx", "Wolf", "Fox", "Coyote", "I don't know"] (randomize order true except for ["Wolf", "Fox", "Other", "I don't know"]): data 1`] = `
Array [
  Array [
    "Lynx",
    true,
  ],
  Array [
    "Panda",
    false,
  ],
  Array [
    "Coyote",
    false,
  ],
  Array [
    "Wolf",
    true,
  ],
  Array [
    "Fox",
    true,
  ],
  Array [
    "Other",
    false,
  ],
  Array [
    "I don't know",
    true,
  ],
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Coyote", "Lynx", "Wolf", "Fox", "Coyote", "I don't know"] (randomize order true except for ["Wolf", "Fox", "Other", "I don't know"]): displayed list 1`] = `
Array [
  "Lynx",
  "Panda",
  "Coyote",
  "Wolf",
  "Fox",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "I don't know"] (randomize order true except for undefined): data 1`] = `
Array [
  Array [
    "Lynx",
    false,
  ],
  Array [
    "I don't know",
    true,
  ],
  Array [
    "Wolf",
    false,
  ],
  Array [
    "Panda",
    false,
  ],
  Array [
    "Other",
    false,
  ],
  Array [
    "Fox",
    false,
  ],
  Array [
    "Coyote",
    false,
  ],
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "I don't know"] (randomize order true except for undefined): displayed list 1`] = `
Array [
  "Lynx",
  "I don't know",
  "Wolf",
  "Panda",
  "Other",
  "Fox",
  "Coyote",
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Fox"] (randomize order true except for ["Other", "I don't know"]): data 1`] = `
Array [
  Array [
    "Lynx",
    false,
  ],
  Array [
    "Coyote",
    false,
  ],
  Array [
    "Wolf",
    false,
  ],
  Array [
    "Panda",
    false,
  ],
  Array [
    "Fox",
    true,
  ],
  Array [
    "Other",
    false,
  ],
  Array [
    "I don't know",
    false,
  ],
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Fox"] (randomize order true except for ["Other", "I don't know"]): displayed list 1`] = `
Array [
  "Lynx",
  "Coyote",
  "Wolf",
  "Panda",
  "Fox",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Lynx", "Panda", "Other"] (randomize order true except for ["Other"]): data 1`] = `
Array [
  Array [
    "Lynx",
    true,
  ],
  Array [
    "Coyote",
    false,
  ],
  Array [
    "Wolf",
    false,
  ],
  Array [
    "Panda",
    true,
  ],
  Array [
    "I don't know",
    false,
  ],
  Array [
    "Fox",
    false,
  ],
  Array [
    "Other",
    true,
  ],
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Lynx", "Panda", "Other"] (randomize order true except for ["Other"]): displayed list 1`] = `
Array [
  "Lynx",
  "Coyote",
  "Wolf",
  "Panda",
  "I don't know",
  "Fox",
  "Other",
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Wolf", "Coyote"] (randomize order false except for undefined): data 1`] = `
Array [
  Array [
    "Wolf",
    true,
  ],
  Array [
    "Fox",
    false,
  ],
  Array [
    "Coyote",
    true,
  ],
  Array [
    "Lynx",
    false,
  ],
  Array [
    "Panda",
    false,
  ],
  Array [
    "Other",
    false,
  ],
  Array [
    "I don't know",
    false,
  ],
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence ["Wolf", "Coyote"] (randomize order false except for undefined): displayed list 1`] = `
Array [
  "Wolf",
  "Fox",
  "Coyote",
  "Lynx",
  "Panda",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence [] (randomize order false except for undefined): data 1`] = `
Array [
  Array [
    "Wolf",
    false,
  ],
  Array [
    "Fox",
    false,
  ],
  Array [
    "Coyote",
    false,
  ],
  Array [
    "Lynx",
    false,
  ],
  Array [
    "Panda",
    false,
  ],
  Array [
    "Other",
    false,
  ],
  Array [
    "I don't know",
    false,
  ],
]
`;

exports[`ChoicesWithMultipleAnswers question with input sequence [] (randomize order false except for undefined): displayed list 1`] = `
Array [
  "Wolf",
  "Fox",
  "Coyote",
  "Lynx",
  "Panda",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Coyote", "Coyote", "Wolf", "Wolf", "Coyote"] (randomize order true except for ["I don't know", "Other"]): data 1`] = `"Coyote"`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Coyote", "Coyote", "Wolf", "Wolf", "Coyote"] (randomize order true except for ["I don't know", "Other"]): displayed list 1`] = `
Array [
  "Lynx",
  "Coyote",
  "Wolf",
  "Panda",
  "Fox",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Coyote", "Lynx", "Wolf", "Fox", "Coyote", "I don't know"] (randomize order true except for ["Wolf", "Fox", "Other", "I don't know"]): data 1`] = `"I don't know"`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Coyote", "Lynx", "Wolf", "Fox", "Coyote", "I don't know"] (randomize order true except for ["Wolf", "Fox", "Other", "I don't know"]): displayed list 1`] = `
Array [
  "Lynx",
  "Panda",
  "Coyote",
  "Wolf",
  "Fox",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "I don't know"] (randomize order true except for undefined): data 1`] = `"I don't know"`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "Coyote", "Wolf", "Fox", "Lynx", "Panda", "Other", "I don't know"] (randomize order true except for undefined): displayed list 1`] = `
Array [
  "Lynx",
  "I don't know",
  "Wolf",
  "Panda",
  "Other",
  "Fox",
  "Coyote",
]
`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Fox"] (randomize order true except for ["Other", "I don't know"]): data 1`] = `"Fox"`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Fox"] (randomize order true except for ["Other", "I don't know"]): displayed list 1`] = `
Array [
  "Lynx",
  "Coyote",
  "Wolf",
  "Panda",
  "Fox",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Lynx", "Panda", "Other"] (randomize order true except for ["Other"]): data 1`] = `"Other"`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Lynx", "Panda", "Other"] (randomize order true except for ["Other"]): displayed list 1`] = `
Array [
  "Lynx",
  "Coyote",
  "Wolf",
  "Panda",
  "I don't know",
  "Fox",
  "Other",
]
`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Wolf", "Coyote"] (randomize order false except for undefined): data 1`] = `"Coyote"`;

exports[`ChoicesWithSingleAnswer question with input sequence ["Wolf", "Coyote"] (randomize order false except for undefined): displayed list 1`] = `
Array [
  "Wolf",
  "Fox",
  "Coyote",
  "Lynx",
  "Panda",
  "Other",
  "I don't know",
]
`;

exports[`ChoicesWithSingleAnswer question with input sequence [] (randomize order false except for undefined): data 1`] = `null`;

exports[`ChoicesWithSingleAnswer question with input sequence [] (randomize order false except for undefined): displayed list 1`] = `
Array [
  "Wolf",
  "Fox",
  "Coyote",
  "Lynx",
  "Panda",
  "Other",
  "I don't know",
]
`;

exports[`YesNo question with input false: displayed list 1`] = `
Array [
  "Yes",
  "No",
]
`;

exports[`YesNo question with input true: displayed list 1`] = `
Array [
  "Yes",
  "No",
]
`;
